docs/src/keymap-upgrade.js

```diff
@@ -0,0 +1,234 @@
+    console.warn(
+      `Found deprecated code "${node.text}" but it is not a parameter to a behavior`
+    );
+    return [new TextEdit(node, `/* "${node.text}" no longer exists */`)];
+  }
+
+  const oldText = nodes.map((n) => n.text).join(" ");
+  const newText = `&none /* "${oldText}" no longer exists */`;
+
+  const startIndex = nodes[0].startIndex;
+  const endIndex = nodes[nodes.length - 1].endIndex;
+
+  return [new TextEdit(startIndex, endIndex, newText)];
+}
+
+/**
+ * Returns the node for the named capture.
+ * @param {string} name
+ * @param {any[]} captures
+ * @returns {Parser.SyntaxNode | null}
+ */
+function findCapture(name, captures) {
+  for (const c of captures) {
+    if (c.name === name) {
+      return c.node;
+    }
+  }
+
+  return null;
+}
+
+/**
+ * Given a parameter to a keymap behavior, returns a list of nodes beginning
+ * with the behavior and including all parameters.
+ * Returns an empty array if no behavior was found.
+ * @param {Parser.SyntaxNode} paramNode
+ */
+function findBehaviorNodes(paramNode) {
+  // Walk backwards from the given parameter to find the behavior reference.
+  let behavior = paramNode.previousNamedSibling;
+  while (behavior && behavior.type !== "reference") {
+    behavior = behavior.previousNamedSibling;
+  }
+
+  if (!behavior) {
+    return [];
+  }
+
+  // Walk forward from the behavior to collect all its parameters.
+
+  let nodes = [behavior];
+  let param = behavior.nextNamedSibling;
+  while (param && param.type !== "reference") {
+    nodes.push(param);
+    param = param.nextNamedSibling;
+  }
+
+  return nodes;
+}
+
+/**
+ * Gets a list of text edits to apply based on a node and a map of text
+ * replacements.
+ *
+ * If replaceHandler is given, it will be called if the node matches a
+ * deprecated value and it should return the text edits to apply.
+ *
+ * @param {Parser.SyntaxNode} node
+ * @param {Map<string, string | null>} replacementMap
+ * @param {(node: Parser.SyntaxNode, replacement: string | null) => TextEdit[]} replaceHandler
+ */
+function getUpgradeEdits(node, replacementMap, replaceHandler = undefined) {
+  for (const [deprecated, replacement] of Object.entries(replacementMap)) {
+    if (node.text === deprecated) {
+      if (replaceHandler) {
+        return replaceHandler(node, replacement);
+      } else {
+        return [new TextEdit(node, replacement)];
+      }
+    }
+  }
+  return [];
+}
+
+/**
+ * Sorts a list of text edits in ascending order by position.
+ * @param {TextEdit[]} edits
+ */
+function sortEdits(edits) {
+  return edits.sort((a, b) => a.startIndex - b.startIndex);
+}
+
+/**
+ * Returns a string with text replacements applied.
+ * @param {string} text
+ * @param {TextEdit[]} edits
+ */
+function applyEdits(text, edits) {
+  edits = sortEdits(edits);
+
+  console.log(edits);
+
+  /** @type string[] */
+  const chunks = [];
+  let currentIndex = 0;
+
+  for (const edit of edits) {
+    if (edit.startIndex < currentIndex) {
```

Comment:
Given something silly like `&cp CURU`, we can both try to change `&cp` to `&kp` and `&cp CURU` to `&none /* "&cp CURU" no longer exists. */`. Those edits overlap so they can't both be applied.

That shouldn't happen in any reasonable keymap so I just made it throw out an arbitrary one of the edits instead of trying to do something smarter to resolve that.